### DISCUSSION TITLE
Linux build guide for Unreal Engine 5.2

### DISCUSSION THREAD

**Discussion by ericotsn** [microdee/OpenUnrealSpaceMouse#91] [Show and tell]
First of all, thank you so much @microdee for your work on this plugin :heart: Support for my SpaceMouse was one of the last things keeping me back from moving my game dev workflow over to Linux.

With that said, it wasn't completely straightforward how to get the plugin working on Linux. But after compiling some notes from older GitHub issues and performing some small tweaks I can now navigate my scenes with ease once again! I'm hoping this guide will help others who are looking into how to use their favorite tools on Linux.

First of all, this is my system configuration with a clean OS install and `spacenavd` (for Blender support) installed:

 - OS: Linux Mint 21.2
 - Engine Version: 5.2.1

## Preparation

Start by cloning the repository including submodules:

```
git clone --recurse-submodules https://github.com/microdee/UE4-SpaceMouse.git
```

Next we need to update the `SpaceMouse.uplugin` file and whitelist Linux as a module platform. For each module in the file (4 in total) make sure `WhitelistPlatforms` looks as follows:

```
"WhitelistPlatforms": [
    "Linux",
    "Win64"
]
```

## Building

You can build the plugin using the `RunUAT.sh` CLI tool included in your engine install.

> When you try to build the plugin you might get an error saying `libudev.h not found`. If that's the case, install the `libudev-dev` package from your package manager.

Open your favorite terminal and paste the following command, substituting with your chosen plugin and engine installation directories:

```
/home/erlook/.local/share/UnrealEngine/Engine/Build/BatchFiles/RunUAT.sh BuildPlugin \
-Plugin="/home/erlook/Documents/UE4-SpaceMouse/SpaceMouse.uplugin" \
-Package="/home/erlook/Documents/UE4-SpaceMouse/Out/Linux"
```

This will build the plugin and make it available in the `/UE4-SpaceMouse/Out/Linux` directory.

## Installation

After this you can copy the `Linux` directory into your UE5.2 project's `Plugins` directory and rename it `SpaceMouse`. Start the Unreal Editor and verify that the plugin is enabled in **Edit > Plugins > Input Devices**. Then check that plugin settings are available in **Edit > Editor Preferences > Plugins > SpaceMouse**.

If everything is there the plugin was built successfully and ready to be used! However, it could be that nothing happens when you try to move using the SpaceMouse within your scene. If so, you might need to add a `udev` rule to help your device communicate with the plugin. Copy the following content and save it in a file named `90-3dconnexion.rules` in the `/usr/lib/udev/rules.d` directory. This will require root privileges!

```
# 3D Connexion vendor devices
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c603", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c605", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c606", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c621", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c623", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c625", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c626", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c627", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c628", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c629", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c62b", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c62e", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c62f", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c631", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c632", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c633", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c635", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c636", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c640", MODE="0666"
KERNEL=="hidraw*", ATTRS{idVendor}=="256f", ATTRS{idProduct}=="c652", MODE="0666"
```

Restart your computer and you should now be able to use your SpaceMouse in Unreal Engine 5.2! :tada:

## References

 - Linux support GitHub issue: https://github.com/microdee/UE4-SpaceMouse/issues/63
 - 3D Connexion device udev rules: https://github.com/prusa3d/PrusaSlicer/blob/master/resources/udev/90-3dconnexion.rules

**Comment by microdee** [comment]
Thank you so much for this guide! ❤️ I'm gonna pin it. If you find some time, can you make a PR where you add this as LINUX.md into the repository link it in README.md and put the whitelisted platforms in the .uplugin?

**Reply by ericotsn** [reply]
Sure! I'll try to have the PR ready this week 😄 I'll add some additional troubleshooting info as well which I didn't include in the guide to keep it short and simple.
